### ISSUE TITLE
Gas Optimisation Possibilities in the Stealth Key Registry Contract

### ISSUE THREAD

**Issue by seresistvanandras** [ScopeLift/umbra-protocol#252]
### **tl;dr:** **we can save one `SSTORE` operation (`20,000` or `5,000` gas) by squeezing the public key prefixes into a single storage slot.**

Maybe, you already thought about optimizing the gas consumption of the stealth key registry contract and then you did abandon this path because it might complicate the logic of the contract a little bit. But, on the positive side, we can save the gas costs of an `SSTORE` operation, which amounts to `20,000` gas for the first time assessing the slot and `5,000` gas for subsequent updates.

Currently, this is the storage layout of the stealth key registry contract. 

`mapping(address => mapping(uint256 => uint256)) keys;`

In the mapping, every address stores the following four variables, each with `uint256` types.

```
uint256 _spendingPubKeyPrefix,
uint256 _spendingPubKey,
uint256 _viewingPubKeyPrefix,
uint256 _viewingPubKey
```
In the end, these four variables are stored in the mapping in the following way.
```
keys[_registrant][_spendingPubKeyPrefix] = _spendingPubKey;
keys[_registrant][_viewingPubKeyPrefix] = _viewingPubKey;
```
Note, that the contract somewhat wastes storage slots, since `_spendingPubKey` and `_viewingPubKey` are treated as `uint256`. However, they don't necessarily need a full-length integer (`256` bits), since they can only have two values: `2` and `3`. These two prefixes could be stored in a single storage slot, allowing users to burn significantly less gas at the expense of slightly more complex smart contract logic.

The new layout of the mapping could be something like this:
`mapping(address => mapping(uint256 =>PubKeys)) keys;`,
where PubKeys is a struct that stores the pair of public keys (the spending and the viewing keys). The key of the inner mapping can conveniently encode in `256` bits **_BOTH_** of the prefixes.

**### Discussion:**

- This gas optimization is only an issue on-chain, if the stealth key registry is deployed on a rollup or a side-chain, then the transactions costs are already pretty low, so maybe one does not want to optimize the storage layout?
- The contract will have a slightly more complex structure. 
- On Layer-1, interacting with the registry (setting and updating the public keys) will become approximately `~25%` cheaper.

Is there any drawback of this apart from the slightly increased complexity of the contract? or was this already discussed on some forum and I might have just missed that discussion?

**Comment by mds1** [comment]
Actually we only use two storage slots currently! 🙂 

One slot is used for the 32 byte `_spendingPubKey`, and one slot for the 32 byte `_viewingPubKey`. The prefixes are never written to storage because the keys for Solidity mappings are never written to storage—they are hashed when reading/writing to compute the slot location, but never stored. So when you have a mapping like `mapping(address => mapping(uint256 => uint256)) keys;`, only the final `uint256 keys` live in storage, but the `address` and `uint256` mapping keys do not

The solidity docs have more info on this [here](https://docs.soliditylang.org/en/latest/internals/layout_in_storage.html#mappings-and-dynamic-arrays)

**Comment by seresistvanandras** [comment]
Woah! I did not know this! Thanks for the lesson!
